File: Python/A_Project/A_Tic_Tac_Toe/GameStateHandler/ConcreteStates/o_turn_state.py
from CentralEnum.symbol import Symbol
from GameStateHandler.game_state import GameState


class OTurnState(GameState):
    def next(self, context, player, has_won: bool):
        if has_won:
            from GameStateHandler.ConcreteStates.o_won_state import OWonState
            from GameStateHandler.ConcreteStates.x_won_state import XWonState

            context.set_state(OWonState() if player.get_symbol() == Symbol.O else XWonState())
        else:
            from GameStateHandler.ConcreteStates.x_turn_state import XTurnState

            context.set_state(XTurnState())

    def is_game_over(self) -> bool:
        return False

    def get_result_message(self) -> str:
        return "Game is still in progress."

File: Ruby/A_Project/A_Tic_Tac_Toe/GameStateHandler/ConcreteStates/o_turn_state.rb
require_relative '../game_state'

module GameStateHandler
  module ConcreteStates
    class OTurnState
      include GameState

      def next(context, player, has_won)
        if has_won
          require_relative './x_won_state'
          require_relative './o_won_state'
          context.state = (player.symbol == CentralEnum::Symbol::O ? OWonState.new : XWonState.new)
        else
          require_relative './x_turn_state'
          context.state = XTurnState.new
        end
      end

      def game_over?
        false
      end

      def result_message
        "Game is still in progress."
      end
    end
  end
end
